### PR TITLE
Add support for op kernel type control required types, require int64 for some ops

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Adds support for required types to the op kernel type control infrastructure. Required types are always enabled.
Added int64 as a required type for certain ops.

**Motivation and Context**
In certain cases, we would like to always enable the op kernel implementation for a certain type.
For example, ops that might reasonably be used in shape-related computations, i.e., those that follow a Shape op which has int64 output, should always have their int64 implementation enabled.

### CODE REVIEW

File: onnxruntime/core/providers/op_kernel_type_control.h [hunk 1]
```diff
@@ -9,22 +9,38 @@
...
- * - Allowed types are the types for which support is requested (for example, by external configuration).
- * - Enabled types are the types that are supported in the actual, compiled implementation. They are obtained from the
- *   intersection of supported and allowed types.
+ * 3. Allowed types are the types for which support is requested (for example, by external configuration).
+ *    Zero or more allowed type sets may be given.
+ *    The supported type set (1) will be limited by set intersection with all allowed type sets.
  *
- * The types are associated with an Op argument. It is also possible to specify a global list of allowed types.
+ * 4. Enabled types are the types that are actually supported.
+ *    These are the required types and the supported, allowed types.
```
Comment (skottmckay) [inline onnxruntime/core/providers/op_kernel_type_control.h:32]: might be clearer to say 'actually supported in this build'

File: onnxruntime/core/providers/op_kernel_type_control.h [hunk 2]
```diff
@@ -238,6 +287,88 @@ struct EnabledTypes {
...
+ *
+ * @param OpProvider The Op provider.
+ * @param OpDomain The Op domain.
+ * @param OpName The Op name.
+ * @param ArgDirection Direction of the given Op kernel argument - Input or Output.
+ * @param ArgIndex Index of the given Op kernel argument.
+ * @param RequiredTypeList The types.
+ */
+#define ORT_SPECIFY_OP_KERNEL_ARG_REQUIRED_TYPE_LIST_ALL_OPSETS(                                  \
+    OpProvider, OpDomain, OpName, ArgDirection, ArgIndex, RequiredTypeList)                       \
```
Comment (skottmckay) [inline onnxruntime/core/providers/op_kernel_type_control.h:328]: Would a required type ever change between opsets?

Comment (edgchen1) [reply onnxruntime/core/providers/op_kernel_type_control.h:328]: One example is Max, where the required int64_t is added in a later opset:
```
ORT_SPECIFY_OP_KERNEL_ARG_SUPPORTED_TYPES(kCpuExecutionProvider, kOnnxDomain, Max, 8, Input, 0, float, double);

ORT_SPECIFY_OP_KERNEL_ARG_SUPPORTED_TYPES(kCpuExecutionProvider, kOnnxDomain, Max, 12, Input, 0,
                                          float, double, MLFloat16, int32_t, uint32_t, int64_t, uint64_t);
ORT_SPECIFY_OP_KERNEL_ARG_REQUIRED_TYPES(kCpuExecutionProvider, kOnnxDomain, Max, 12, Input, 0,
                                         int64_t);
```

File: onnxruntime/core/providers/op_kernel_type_control.h (hunk 1 above)
Comment (edgchen1) [reply onnxruntime/core/providers/op_kernel_type_control.h:32]: ok. it occurred to me that 'supported' is overloaded in this description - i think i will rename 'supported types' to 'default types' in a future change.

**Review by skottmckay — APPROVED** [review]

**Pull request merged**